trusted_applet/internal/update/update.go

```diff
@@ -0,0 +1,131 @@
+
+	// InstallApplet updates the Applet to that included in the firmware bundle.
+	InstallApplet(firmware.Bundle) error
+
+	// Reboot instructs the device to reboot after new firmware is installed.
+	// This call will not return.
+	Reboot()
+}
+
+// A Remote represents the connection to the Internet and allows access to
+// query and fetch new versions of firmware.
+type Remote interface {
+	// GetLatestVersions returns the latest available versions of the OS and Applet.
+	GetLatestVersions() (os, applet semver.Version, err error)
+	// GetOS fetches the operating system executable and associated metadata.
+	GetOS() (firmware.Bundle, error)
+	// GetApplet fetches the applet executable and associated metadata.
+	GetApplet() (firmware.Bundle, error)
+}
+
+// A FirmwareVerifier checks that the given Bundle passes installation policy.
+type FirmwareVerifier interface {
+	// Verify checks the firmware bundle and returns an error if invalid, or nil
+	// if the firmware is safe to install.
+	Verify(firmware.Bundle) error
+}
+
+// NewUpdater returns an Updater that uses local to query/update the device, remote to
+// query/fetch new updates, and verifier to ensure that downloaded content passes installation
+// policy.
+func NewUpdater(local Local, remote Remote, verifier FirmwareVerifier) (*Updater, error) {
+	osVer, appVer, err := local.GetInstalledVersions()
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine installed versions: %v", err)
+	}
+	return &Updater{
+		local:    local,
+		remote:   remote,
+		verifier: verifier,
+		osVer:    osVer,
+		appVer:   appVer,
+	}, nil
+}
+
+// Updater should be periodically invoked via Update to check for and install firmware
+// udpates for the OS and applet.
+type Updater struct {
+	local         Local
+	remote        Remote
+	verifier      FirmwareVerifier
+	osVer, appVer semver.Version
+}
+
+// Update checks whether newer versions of installed firmware are available, and if so
+// it fetches, verifies, and installs the firmware. If any changes are made to the
+// firmware on disk then this function will not return because a reboot will be scheduled.
+// This function is designed to be called periodically by a single thread. It is not
+// thread safe.
+func (u Updater) Update(ctx context.Context) error {
+	osVer, appVer, err := u.remote.GetLatestVersions()
+	if err != nil {
+		return fmt.Errorf("failed to get latest versions: %v", err)
+	}
+	if u.osVer.LessThan(osVer) {
+		glog.Infof("Upgrading OS from %q to %q", u.osVer, osVer)
+		bundle, err := u.remote.GetOS()
+		if err != nil {
+			return fmt.Errorf("failed to fetch OS firmware: %v", err)
+		}
+		if err := u.verifier.Verify(bundle); err != nil {
+			return fmt.Errorf("verification of OS firmware bundle failed: %v", err)
+		}
+		if err := u.local.InstallOS(bundle); err != nil {
+			return fmt.Errorf("failed to install OS firmware: %v", err)
+		}
+		// Defer rebooting until we've attempted updating the Applet
+		defer u.local.Reboot()
```

Comment:
If both the OS and applet are updated, the device will be rebooted twice. I think it's cleaner to factor this out to a final if block at the end to only reboot once. WDYT?

Reply:
It won't really reboot twice because the Reboot docs describe that it won't return. So only one or the other will execute. I've added to the Reboot docs to borrow from os.Exit to say that deferred functions will not be run.

I understand the concern though and considered this when writing. I believe that this is safe and any other alternative is actually more prone to error. In particular, anything other than a defer requires work in all the cases below in which we return an error.